### ISSUE TITLE
Rename OMZ's master branch

### ISSUE THREAD

**Issue by robbyrussell** [ohmyzsh/ohmyzsh#9015]
We are going rename our primary/main branch and discontinue using `master`.

Proposal to move from `main` (or `stable` as we're also exploring a `beta` branch naming convention for community-driven testing).  @mcornella can help make a call there.

Given that our install and update process is reliant on branch names, this will not be a quick nor simple change. ...but it will (and needs to) happen. Thanks in advance for your contributions and patience as we figure out a strategy.

* * * *

```[tasklist]
### Tasks
- [ ] Add logic to `upgrade.sh` to locally rename the branch
- [ ] Test branch rename on external fork
- [ ] Test branch rename against zsh plugin managers:
	- [ ] https://github.com/zsh-users/antigen
	- [ ] https://github.com/getantibody/antibody
	- [ ] https://getantidote.github.io/
	- [ ] https://github.com/tarjoilija/zgen
	- [ ] https://github.com/zdharma-continuum/zinit
	- [ ] https://github.com/zplug/zplug
	- [ ] https://github.com/zpm-zsh/zpm
- [ ] Test branch rename against a fork
- [ ] Rename in GitHub UI
```

**Comment by mcornella** [comment]
On a second thought, maybe we can still use the `master` but as the beta/edge branch, and use tagged commits in the master branch for stable releases. This simplifies maintenance a great deal, because we don't have to deal with merges and backports from one branch to the other.

Another advantage of using git tags is that we can use the annotated tags feature to add a changelog. Of course, this also implies changing the update system but it's a lot simpler.

Another thing we discussed is the naming of versions. Since OMZ is mainly a rolling release project, it's difficult to decide on a versioning scheme. Some of the options:
- [Semantic versioning](https://semver.org/): `v<major>.<minor>.<patch>`.
- Rev count: `v<rev-count>`, where `<rev-count>` is the number of commits.
- A mix of both, preserving the `v<major>` prefix and adding `<rev-count>`, because it might not make sense to distinguish between `<minor>` and `<patch>` increments.

It's important to note that this is still a discussion in progress.

**Comment by robbyrussell** [comment]
@mcornella Apologies for the lack of clarity. My goal here was to discontinue naming any branch in the project 'master'. 

i.e., https://www.hanselman.com/blog/EasilyRenameYourGitDefaultBranchFromMasterToMain.aspx

**Comment by mcornella** [comment]
Oh sorry, I got confused because we were also talking about versioning. This looks reasonable and a step in the right direction. I'll have a look at needed migration steps.

**Comment by drizzt** [comment]
I think it's "slave" the problematic word. The origin of word master is "magister" that means teacher.

**Comment by robbyrussell** [comment]
@drizzt Cool. Feel free to rename your project branches teacher. We are moving in a different direction.

**Comment by adam2k** [comment]
@robbyrussell does this also include changing the `git` plugin's command `gcm` to be an alias for `git checkout main`?  If not, I can create a new issue and PR for that.

**Comment by askpatrickw** [comment]
> @robbyrussell does this also include changing the `git` plugin's command `gcm` to be an alias for `git checkout main`? If not, I can create a new issue and PR for that.

I came here looking for this exact thing. 

I do believe gcm should be main. Unless GitHub picks some other default of course.
Maybe having a GCM variable to set it to your preference would be good as well as not all repos can change or will change to the same thing.

**Comment by mcornella** [comment]
@askpatrickw does this PR solve your needs? https://github.com/ohmyzsh/ohmyzsh/pull/9049

**Comment by askpatrickw** [comment]
@mcornella I believe it will, thank you !

**Comment by ljmf00** [comment]
Hi @robbyrussell ,

I want to clarify something about this topic:

1. See the definition of `master` branch as ['master copy'][1] and understand why it's named that way. In `git`, other branches rather than `master` branch, usually share the same history with `master` and that's why `master` is called that way. Please understand and accept that `master` branch is nothing to do with `master`/`slave` principle and no one feel offended except you and who think that changing the name of `master` branch on git and other huge protocols that has `master` on it will solve racism/inequality/discrimination in this world.

2. This will cause a huge breaking change. Other repositories that depends on this one will rely on `master` branch, because, unfortunately, there's no tagging on OMZ repository, see #9216 . Even if you say, no one depends on this repository, the update script that is used by `oh-my-zsh` itself, is relying on the current user's default branch, which is `master`. Remember, this script is present on EVERY single `/home/` directory of whatever user is using OMZ and surely it will break the users environment.

3. OMZ has 200+ active issues and growing, can we do better? There's a lot more important stuff to discuss and deal with...

In conclusion: don't break software and the user's environment. A lot of people who reacted negatively to your issue description, may understand the point described here.

Also:
> @drizzt Cool. Feel free to rename your project branches teacher. We are moving in a different direction.

IMHO, your comment on this is just unnecessary.


[1]: https://www.collinsdictionary.com/dictionary/english/master-copy

**Comment by abraham** [comment]
> Please understand and accept that master branch is nothing to do with `master`/`slave` principle

This is incorrect. The term `master` in git is almost certainly based on the `master`/`slave` meaning from BitKeeper.

https://english.stackexchange.com/a/539969/90923

**Comment by ljmf00** [comment]
> Git branches have been qualified as topic branches, integration branches,
development branches, feature branches, release branches and so on.
Git has a branch that is the master *for* development, but it is not
the master *of* any "slave branch": Git does not have slave branches,
and has never had, except for a single testcase that claims otherwise. :)

Source: https://github.com/git/git/commit/08dc26061f3ff9ee79e6cfda88f0c825b8730e54

---

Searching the "slave" term on the default branch, you will find [this](https://github.com/git/git/search?q=slave&unscoped_q=slave), and if you do a simple `git blame` on https://github.com/git/git/blob/53f9a3e157dbbc901a02ac2c73346d375e24978c/t/test-terminal.perl, you see that it's a simple testcase and was not introduced by the major active developers of `git`.

> This is incorrect. The term master in git is almost certainly based on the master/slave meaning from BitKeeper.

So, even if git was based on BitKeeper and the word `master` is there because of that, doesn't necessarily mean that they went or are going in that direction.

My major point is not being against the word, is more about the side effects of changing the name of the default branch. Honestly, `main` could be a better word to describe the purpose, but `git` just named that way.

**Comment by mikestopcontinues** [comment]
The word master pre-dates its usage in the US plantation system [by thousands of years](https://www.lexico.com/en/definition/master), and the word is still used in many other capacities. This move re-racializing Western society by increasing the number of taboos around skin tone. Better that we allow people of every color to manage _master_ branches, administrate _master/slave_ database clusters, reside in the _master_ bedroom, and _master_ a wide range of skills. Here's the titanic (and black and liberal) linguist John McWhorter [on the issue](https://reason.com/2020/08/12/is-your-master-bedroom-racist/).

**Comment by mcornella** [comment]
As Robby said previously, this is happening without a doubt. The issue remains to sort out the technical issues to be aware. It _is_ possible to implement the rename without problems, it has been done in other projects and it will be done in this one as well, and no user will even notice.

**Comment by Samega7Cattac** [comment]
I can't believe in what I'm reading

**Comment by ljmf00** [comment]
> As Robby said previously, this is happening without a doubt.

This is just sad. You guys literally state that OMZ is a "community-driven" project and you make decisions without caring about who uses your software and the stability of it for the community around, even if a more than half of the people that reacts to this issue are completely upset with the decision.

> It is possible to implement the rename without problems, it has been done in other projects and it will be done in this one as well

No it's not. I'm sorry, but you will always have a problem here. There's no other reference rather than `refs/heads/master` at the moment and as long as you run `oh-my-zsh` for years on your system with this https://github.com/ohmyzsh/ohmyzsh/blob/master/tools/upgrade.sh#L58 piece of code, you will end up breaking the users environment, because guess what, the only official way to use `oh-my-zsh` is by depending on `git` to fetch updates from this repository.

Why don't you just care more about other existing problems in the project instead of creating new ones?

**Comment by mcornella** [comment]
> There's no other reference rather than `refs/heads/master` at the moment and as long as you run `oh-my-zsh` for years on your system with this [https://github.com/ohmyzsh/ohmyzsh/blob/master/tools/upgrade.sh#L58](https://github.com/ohmyzsh/ohmyzsh/blob/master/tools/upgrade.sh?rgh-link-date=2020-09-21T11%3A18%3A50Z#L58) piece of code, you will end up breaking the users environment, because guess what, the only official way to use `oh-my-zsh` is by depending on `git` to fetch updates from this repository.

This is entirely changeable. The current thinking is to maintain both branches side by side and pointing both to the same commit, then fix the upgrade script to use the main branch instead. A similar algo was written to update the remote URL when migrating from @robbyrussell's user to the ohmyzsh organization.

After that is in place, we will update the default branch to main via the GitHub settings UI, and progressively move all the software that uses Oh My Zsh to use `main`. This includes zsh package managers like antigen, zplug, zinit and others.

Also, we will start tagging commits that have been tested enough with version numbers so that maintainers that package ohmyzsh will have an better experience.

All of these changes contribute to improving the overall Oh My Zsh experience and we'd do them even if we didn't think about renaming the default branch.

> Why don't you just care more about other existing problems in the project instead of creating new ones?

You're welcome to chip in with the existing problems of the project. We're already doing it.

**Comment by ljmf00** [comment]
> This is entirely changeable. The current thinking is to maintain both branches side by side and pointing both to the same commit, then fix the upgrade script to use the main branch instead.

I'm fine if you don't delete `refs/heads/master` branch, but if you do it afterwards, you will break users environment anyway.

> A similar algo was written to update the remote URL when migrating from @robbyrussell's user to the ohmyzsh organization.

I'm not aware of that change, but it seems that @robbyrussell remote was deleted, but fortunately, Github does have an alias for you to safely redirect to this repo, otherwise, it would be a breaking change, for sure.

> Also, we will start tagging commits that have been tested enough with version numbers so that maintainers that package ohmyzsh will have an better experience.

So, please, track the progress on the issue I created about version tagging. I suggested using https://semver.org/ .

> You're welcome to chip in with the existing problems of the project. We're already doing it.

You didn't get my point... Instead of potentially create a breaking change because of a naming convention and wasting a lot of time on this, why not focus on the existing **real** problems?

Anyway, my time is not unlimited and I see no reason to continue discussing pointless issues, especially when you don't care about the opinion of people and say: "this is happening without a doubt".

**Comment by mcornella** [comment]
> I'm not aware of that change, but it seems that @robbyrussell remote was deleted, but fortunately, Github does have an alias for you to safely redirect to this repo, otherwise, it would be a breaking change, for sure.

This is the [algorithm which made the remote change](https://github.com/ohmyzsh/ohmyzsh/blob/93c837fec8e9fe61509b9dff9e909e84f7ebe32d/tools/upgrade.sh#L51-L55). We can apply a similar solution to change user's ohmyzsh folders to checkout main instead of master.

My point is that the potential technical issues that might arise from the change are being discussed, and we won't proceed to remove the master branch until we're 100% sure that everybody switched over to main, for example by tracking recent fetches of the master branch.

Removing the master branch is the last of the things to do and something we don't need to do until it doesn't cause any problems, as I said. We can already start implementing the other stuff as soon as we have a roadmap set in stone.

**Comment by mcornella** [comment]
It also seems like [GitHub is working on a tool](https://github.com/github/renaming#later-this-year-seamless-move-for-existing-repositories-) to make the renaming easier so we won't have to wait too much.

**Comment by raxetul** [comment]
`staging` for last testing before mainstream launch (merging into `main` branch), `development` as PR target branch for us, "The Smurfs"

**Comment by Cyberbeni** [comment]
> Removing the master branch is the last of the things to do and something we don't need to do until it doesn't cause any problems, as I said. We can already start implementing the other stuff as soon as we have a roadmap set in stone.

I don't think I'm the only one with their own git repo containing a script using the currently recommended installation method (or even the one pointing to robbyrussel's repository) to setup their workspace whenever they move to a new Mac. So if the current installation method stops working, it WILL break something, no matter how long we wait.

Also never seen anyone who could be referred as "salve" by racist people saying that renaming the default branch from "master" to anything else would help them in any way. Making meaningless changes just detracts from the movement: https://twitter.com/bocxtop/status/1277278544134111239

**Comment by mcornella** [comment]
The installation script won't stop working.

**Comment by ljmf00** [comment]
> The installation script won't stop working.

**cof** *cof*

**Comment by flexbox** [comment]
Hey! 

I landed on this thread because I got some deploy issues with the `main` / `master` drama. 😅

From my persective, I am using shorcurts all the time for ages.
They are really wired to my brain now.

I just wanted to know if you have planned to change the shortcut as well? 
-- | --
gcm | git checkout master
-- | --

**Comment by mcornella** [comment]
Hey @flexbox, the `gcm` alias was already changed to prefer `main` if the branch exists, otherwise use `master`.

**Comment by flexbox** [comment]
> Hey @flexbox, the `gcm` alias was already changed to prefer `main` if the branch exists, otherwise use `master`.

Didn't know 
Thank you @mcornella you are a legend 🚀

**Comment by carmenjohnson992** [comment]
@mcornella @robbyrussell @drizzt @adam2k @askpatrickw @ljmf00 @Cyberbeni @flexbox @hllewelyn @fbender

Wоuld love to test [SITE](https://is.gd/oxMW1N), but I need some tokens. Anyоne generous hеre? 0x2143F7146F0AadC0F9d85ea98F23273Da0e002Ab

**Comment by robbyrussell** [comment]
This was something that I requested, and while we tried to find a streamlined way to do make this change... given that we opted to have the installation be a git clone, figuring out how to make this work nicely for everyone who already installed it felt like far more effort than we felt necessary.

**Comment by robbyrussell** [comment]
I closed this back in January because it had been sitting around for years, and I still couldn't see a clean path forward that didn't risk creating problems for a whole lot of existing installs.

Now that it's easier to bounce these sorts of weird edge-case challenges off an LLM, I spent a little time working through it again with Claude. It came back with an approach that makes me wonder if there might finally be a reasonably safe path forward.

The short version: **don't rename anything.**

Add `main`, make it the default, and keep `master` around indefinitely as a mirror that gets fast-forwarded on every push.

Existing installs could keep pulling from `master` exactly as they do today and never need to know anything changed. Meanwhile, the updater could gradually migrate people over to `main` when it's safe to do so... with every failure path effectively being, "okay, just keep using `master`."

What has always made me nervous about a straightforward rename is that every existing install runs something along the lines of:

```sh
git pull --rebase origin master
```

...with the branch name baked into `oh-my-zsh.branch`.

GitHub can redirect URLs after a branch rename, but that doesn't help an existing clone trying to fetch a ref that no longer exists. With millions of Oh My Zsh installs out there, even a very small percentage of weird old clones, forks, custom setups, or updater edge cases can turn into a pretty big mess.

Keeping `master` around as a mirror seems like it could sidestep most of that.

@mcornella and Carlo, would you be willing to help vet this approach and try to poke holes in it? In particular, I'm curious whether you see problems around the updater, plugin managers, forks, old clones, custom branch settings, or anything else I'm overlooking.

If we can convince ourselves that this gets us reasonably close to zero impact for existing installations, maybe this is something we could actually work toward over the next few months rather than leaving it in the "too risky to touch" pile forever.

## Possible implementation plan

### Core idea

1. Create `main` from `master` and switch the GitHub default branch through the repository settings rather than using GitHub's rename functionality.
2. Add a workflow on pushes to `main` that runs:

   ```sh
   git push origin HEAD:refs/heads/master
   ```

   Plain push, never force. If it can't fast-forward, it should fail loudly.
3. Change `install.sh` to:

   ```sh
   BRANCH=${BRANCH:-main}
   ```

   ...so new installs land on `main`.
4. Add a migration path to `upgrade.sh` that moves existing installs from `master` to `main` when we're confident it's safe.
5. Never delete `master`.

### Migration in `upgrade.sh`

Only attempt the migration when all of these are true:

- `oh-my-zsh.branch` is unset or set to `master`
- local `master` exists and local `main` does not
- we're not in the middle of a rebase or merge
- `git ls-remote --exit-code --heads $remote main` succeeds, which should also protect forks that don't have `main`

Then:

1. Fetch `$remote/main`.
2. Rename the local `master` branch to `main`.
3. Set its upstream to `$remote/main`.
4. Write `oh-my-zsh.branch main`.
5. Print a short message explaining that the branch changed.

If any part of that fails, undo what we can and continue using `master`.

One wrinkle: the updater runs the existing `upgrade.sh` in order to fetch the new one, so the migration would naturally happen one update after the migration code ships. That seems fine.

### Rollout

1. Merge the mirror workflow while `master` is still the default and confirm the two branches stay in sync.
2. Switch the default branch to `main`, move branch protection, and retarget open PRs.
3. Update `install.sh`.
4. Ship the `upgrade.sh` migration along with README/wiki updates and a changelog note.
5. Watch for issues for a few weeks.
6. Leave `master` alone indefinitely.

### Testing

Before shipping the migration, we could script this against a local bare repo standing in for GitHub and run through a bunch of scenarios:

- shallow clone
- full clone
- local edits
- local commits
- existing local `main`
- custom `oh-my-zsh.branch`
- remote without `main`
- detached HEAD
- dirty rebase/merge state

For each one, assert that the branch, config, and upstream end up where we expect... and that running the migration again is a no-op.

_For transparency: I worked through this possible approach collaboratively with Claude Code… so consider the implementation details above a starting point to poke holes in, not a battle-tested plan._

**Comment by night-jellyfish** [comment]
Thanks for re-opening! I did want to respond to these comments:

> Ultimately, the branch name hasn’t been a blocker for millions of people installing and using Oh My Zsh. [...] And if a legacy branch name means someone doesn’t want to recommend Oh My Zsh, that’s their choice.

I think the inclusivity of our community is very important to the growth and sustainability of said community, and I think we don't see the damage done by ignoring these issues. I don't think the people who are comfortable ignoring this issue should be prioritized over the invisible-to-many damage perpetuated by the naming. All that it takes for evil to succeed, first they came for, etc etc.

**Comment by HeyBanditoz** [comment]
What
